File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/dialog_loading/LoadingDialogBuilder.java
package com.github.rooneyandshadows.lightbulb.dialogs.dialog_loading;

import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment;

import androidx.fragment.app.FragmentManager;

public class LoadingDialogBuilder extends BaseDialogBuilder<LoadingDialog> {

    public LoadingDialogBuilder(FragmentManager manager, String dialogTag) {
        super(manager, dialogTag);
    }

    @Override
    public LoadingDialogBuilder withTitle(String title) {
        return (LoadingDialogBuilder) super.withTitle(title);
    }

    @Override
    public LoadingDialogBuilder withMessage(String message) {
        return (LoadingDialogBuilder) super.withMessage(message);
    }

    @Override
    public LoadingDialogBuilder withPositiveButton(BaseDialogFragment.DialogButtonConfiguration positiveButtonConfiguration, BaseDialogFragment.DialogButtonClickListener onClickListener) {
        return (LoadingDialogBuilder) super.withPositiveButton(positiveButtonConfiguration, onClickListener);
    }

    @Override
    public LoadingDialogBuilder withNegativeButton(BaseDialogFragment.DialogButtonConfiguration negativeButtonConfiguration, BaseDialogFragment.DialogButtonClickListener onClickListener) {
        return (LoadingDialogBuilder) super.withNegativeButton(negativeButtonConfiguration, onClickListener);
    }

    @Override
    public LoadingDialogBuilder withOnCancelListener(BaseDialogFragment.DialogCancelListener listener) {
        return (LoadingDialogBuilder) super.withOnCancelListener(listener);
    }

    @Override
    public LoadingDialogBuilder withOnShowListener(BaseDialogFragment.DialogShowListener listener) {
        return (LoadingDialogBuilder) super.withOnShowListener(listener);
    }

    @Override
    public LoadingDialogBuilder withOnHideListener(BaseDialogFragment.DialogHideListener listener) {
        return (LoadingDialogBuilder) super.withOnHideListener(listener);
    }

    @Override
    public LoadingDialogBuilder withCancelOnClickOutsude(boolean closeOnClickOutside) {
        return (LoadingDialogBuilder) super.withCancelOnClickOutsude(closeOnClickOutside);
    }

    @Override
    public LoadingDialogBuilder withDialogType(BaseDialogFragment.DialogTypes dialogType) {
        return (LoadingDialogBuilder) super.withDialogType(dialogType);
    }

    @Override
    public LoadingDialogBuilder withAnimations(BaseDialogFragment.DialogAnimationTypes animation) {
        return (LoadingDialogBuilder) super.withAnimations(animation);
    }

    @Override
    public LoadingDialog buildDialog() {
        LoadingDialog dialogFragment = (LoadingDialog) fragmentManager.findFragmentByTag(dialogTag);
        if (dialogFragment == null)
            dialogFragment = LoadingDialog.newInstance(title, message, dialogType, animation);
        dialogFragment.setFragmentManager(fragmentManager);
        dialogFragment.setDialogTag(dialogTag);
        dialogFragment.addOnShowListener(onShowListener);
        dialogFragment.addOnHideListener(onHideListener);
        dialogFragment.addOnCancelListener(onCancelListener);
        return dialogFragment;
    }
}

File: lightbulb-dialogs/src/main/java/com/github/rooneyandshadows/lightbulb/dialogs/dialog_loading/LoadingDialogBuilder.kt
package com.github.rooneyandshadows.lightbulb.dialogs.dialog_loading

import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.LifecycleOwner
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogAnimationTypes
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogButton
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogTypes
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.callbacks.*

class LoadingDialogBuilder @JvmOverloads constructor(
    dialogTag: String,
    dialogParentFragmentManager: FragmentManager,
    dialogLifecycleOwner: LifecycleOwner? = null,
    initialDialogState: Bundle? = null
) : BaseDialogBuilder<LoadingDialog>(dialogTag, dialogParentFragmentManager, dialogLifecycleOwner, initialDialogState) {

    @JvmOverloads
    constructor(dialogTag: String, fragment: Fragment, initialDialogState: Bundle? = null) : this(
        dialogTag,
        fragment.childFragmentManager,
        fragment,
        initialDialogState
    )

    @JvmOverloads
    constructor(dialogTag: String, activity: FragmentActivity, initialDialogState: Bundle? = null) : this(
        dialogTag,
        activity.supportFragmentManager,
        activity,
        initialDialogState
    )

    @Override
    override fun setupNonRetainableSettings(dialog: LoadingDialog) {
    }

    @Override
    override fun setupRetainableSettings(dialog: LoadingDialog) {
    }

    @Override
    override fun initializeNewDialog(): LoadingDialog {
        return LoadingDialog.newInstance()
    }

    @Override
    override fun withTitle(title: String): LoadingDialogBuilder {
        return super.withTitle(title) as LoadingDialogBuilder
    }

    @Override
    override fun withMessage(message: String): LoadingDialogBuilder {
        return super.withMessage(message) as LoadingDialogBuilder
    }

    @Override
    override fun withButton(configuration: DialogButton): LoadingDialogBuilder {
        return super.withButton(configuration) as LoadingDialogBuilder
    }

    @Override
    override fun withOnCancelListener(listener: DialogCancelListener): LoadingDialogBuilder {
        return super.withOnCancelListener(listener) as LoadingDialogBuilder
    }

    @Override
    override fun withOnShowListener(listener: DialogShowListener): LoadingDialogBuilder {
        return super.withOnShowListener(listener) as LoadingDialogBuilder
    }

    @Override
    override fun withOnHideListener(listener: DialogHideListener): LoadingDialogBuilder {
        return super.withOnHideListener(listener) as LoadingDialogBuilder
    }

    @Override
    override fun withCancelOnClickOutside(closeOnClickOutside: Boolean): LoadingDialogBuilder {
        return super.withCancelOnClickOutside(closeOnClickOutside) as LoadingDialogBuilder
    }

    @Override
    override fun withDialogType(dialogType: DialogTypes): LoadingDialogBuilder {
        return super.withDialogType(dialogType) as LoadingDialogBuilder
    }

    @Override
    override fun withAnimations(animation: DialogAnimationTypes): LoadingDialogBuilder {
        return super.withAnimations(animation) as LoadingDialogBuilder
    }

    @Override
    override fun withDialogListeners(listeners: DialogListeners): LoadingDialogBuilder {
        return super.withDialogListeners(listeners) as LoadingDialogBuilder
    }
}